osu.Game/Scoring/ScorePerformanceManager.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Mods;
+
+namespace osu.Game.Scoring
+{
+    public class ScorePerformanceManager : Component
+    {
+        private readonly ConcurrentDictionary<PerformanceCacheLookup, double> performanceCache = new ConcurrentDictionary<PerformanceCacheLookup, double>();
+
+        [Resolved]
+        private BeatmapManager beatmapManager { get; set; }
+
+        [Resolved]
+        private BeatmapDifficultyManager difficultyManager { get; set; }
+
+        /// <summary>
+        /// Calculates performance for the given <see cref="ScoreInfo"/>.
+        /// </summary>
+        /// <param name="score">The score to do the calculation on. </param>
+        /// <param name="token">An optional <see cref="CancellationToken"/> to cancel the operation.</param>
+        public async Task<double> CalculatePerformanceAsync([NotNull] ScoreInfo score, CancellationToken token = default)
+        {
+            if (tryGetExisting(score, out var perf, out var lookupKey))
+                return perf;
+
+            return await computePerformanceAsync(score, lookupKey, token);
+        }
+
+        private bool tryGetExisting(ScoreInfo score, out double performance, out PerformanceCacheLookup lookupKey)
+        {
+            lookupKey = new PerformanceCacheLookup(score);
+
+            return performanceCache.TryGetValue(lookupKey, out performance);
+        }
+
+        private async Task<double> computePerformanceAsync(ScoreInfo score, PerformanceCacheLookup lookupKey, CancellationToken token = default)
+        {
+            var beatmap = beatmapManager.GetWorkingBeatmap(score.Beatmap);
+            var attributes = await difficultyManager.GetDifficultyAsync(score.Beatmap, score.Ruleset, score.Mods, token);
+
+            if (token.IsCancellationRequested)
+                return default;
+
+            var calculator = score.Ruleset.CreateInstance().CreatePerformanceCalculator(beatmap, score, attributes.Attributes);
+            var total = calculator.Calculate();
+
+            performanceCache[lookupKey] = total;
+
+            return total;
+        }
+
+        public readonly struct PerformanceCacheLookup
+        {
+            public readonly double Accuracy;
+            public readonly int BeatmapId;
+            public readonly long TotalScore;
+            public readonly int Combo;
+            public readonly Mod[] Mods;
+            public readonly int RulesetId;
```

Comment:
This cache lookup is weird. It doesn't use the actual unique-ish attributes of a score (local ID, online ID, user ID, score hash), but instead uses weird heuristics like accuracy or total score. I can guarantee this will cause a mis-hit *somewhere*.

Reply:
Furthermore, you can't use the mods directly. It needs to be acronym and sorted in some fashion - check how `BeatmapDifficultyManager` does its caching.

Reply:
This means that in the case of calculating performance for two exact same scores (same accuracy, total score, combo, same mod combination) on same beatmap using the same ruleset, the performance will be computed twice? Goal I had with including all these values was to cache performance calculation for scores with identical stats.

Reply:
The caching won't be *perfect* with the fields I suggested, yes. The only way to get perfect caching would be to create a cache key based on all inputs to the performance calculator, but given that diffcalc differs from ruleset to ruleset doing so is probably infeasible and definitely overengineering. The attributes you chose *sort of* achieve uniqueness if you're lucky, but not necessarily (imagine two scores with the same accuracy and total score, but e.g. with PERFECTs and OKs swapped - you could mistakenly use an inapplicable cache key and return a wrong value). Score unique attributes are way safer in that regard.